crypto/sha/keccak1600.c

```diff
@@ -11,8 +11,67 @@
 #include <string.h>
 #include <assert.h>
 
-#define ROL64(a, offset) ((offset) ? (((a) << offset) | ((a) >> (64-offset))) \
-                                   : a)
+#define ROL32(a, offset) (((a) << (offset)) | ((a) >> ((32 - (offset)) & 31)))
+
+static uint64_t ROL64(uint64_t val, int offset)
+{
+    if (offset == 0) {
+        return val;
+    } else if (sizeof(void *) == 8) {
```

Comment:
@dot-asm: x86-64 with [the x32 ABI](https://en.wikipedia.org/wiki/X32_ABI) has 32-bit pointers but 64-bit registers, so 64-bit rotates are still efficient.  IDK if there are any other architectures in the same boat, like maybe an ILP32 for ARMv8 ABI that saves memory/cache footprint for pointer-heavy data structures.

x32 hasn't really caught on, so it's still a very niche compiler target at this point, so maybe you don't want to clutter the code with a special case for it.

I didn't check whether gcc/clang still "see through" the code in the other branch of the if and compile it back into a 64-bit rotate, but if not then maybe

```
#ifdef __x86_64__
#  define HAVE_64BIT_REGS 1  // x86-64's x32 ABI uses 32-bit pointers in 64-bit mode
#else
#  define HAVE_64BIT_REGS (sizeof(void*) >= 8)
#endif
```

I don't see a cleaner way to detect a 64-bit machine using 32-bit pointers.  Checking `(sizeof(long) >= 8 || sizeof(void*) >= 8)` doesn't help because x32 is ILP32,   Although that might be a good idea for the non-x32 case in case some other ABIs for 64-bit machines have 64-bit long but 32-bit pointers.

x32's `int_fast32_t` is 32-bit (unlike the normal x86-64 SystemV ABI where it's 64-bit, presumably to avoid sign-extension instructions when used as an array index).